File: www/src/stats.js

export class Stats {
  constructor() {
    this.total = 0;
    this.complete = 0;
    this.incomplete = 0;
    this.citations = 0;
    this.implications = 0;
    this.tests = 0;
    this.exceptions = 0;
    this.todos = 0;
  }

  onRequirement(requirement) {
    this.total += 1;

    if (requirement.incomplete) this.incomplete += 1;
    else if (requirement.isOk) this.complete += 1;

    if (requirement.citation) this.citations += 1;
    if (requirement.implication) this.implications += 1;
    if (requirement.test) this.tests += 1;
    if (requirement.exception) this.exceptions += 1;
    if (requirement.todo) this.todos += 1;
  }

  onStat(stat) {
    this.total += stat.total;
    this.complete += stat.complete;
    this.incomplete += stat.incomplete;
    this.citations += stat.citations;
    this.implications += stat.implications;
    this.tests += stat.tests;
    this.exceptions += stat.exceptions;
    this.todos += stat.todos;
  }

  percent(field) {
    const percent = this.total ? this[field] / this.total : 0;
    return Number(percent).toLocaleString(undefined, {
      style: "percent",
      minimumFractionDigits: 0,
      maximumFractionDigits: 2,
    });
  }
}

File: duvet/src/report/stats.rs
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::Reference;
use crate::annotation::{AnnotationLevel, AnnotationType};

#[derive(Clone, Copy, Debug, Default)]
pub struct Statistics {
    pub must: AnnotationStatistics,
    pub should: AnnotationStatistics,
    pub may: AnnotationStatistics,
}

impl Statistics {
    #[allow(dead_code)]
    pub(super) fn record(&mut self, reference: &Reference) {
        match reference.annotation.level {
            AnnotationLevel::Auto => {
                // don't record auto references
            }
            AnnotationLevel::Must => {
                self.must.record(reference);
            }
            AnnotationLevel::Should => {
                self.should.record(reference);
            }
            AnnotationLevel::May => {
                self.may.record(reference);
            }
        }
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct AnnotationStatistics {
    pub total: Stat,
    pub citations: Stat,
    pub tests: Stat,
    pub exceptions: Stat,
    pub todos: Stat,
    pub implications: Stat,
}

impl AnnotationStatistics {
    #[allow(dead_code)]
    fn record(&mut self, reference: &Reference) {
        self.total.record(reference);
        match reference.annotation.anno {
            AnnotationType::Citation => {
                self.citations.record(reference);
            }
            AnnotationType::Test => {
                self.tests.record(reference);
            }
            AnnotationType::Exception => {
                self.exceptions.record(reference);
            }
            AnnotationType::Todo => {
                self.todos.record(reference);
            }
            AnnotationType::Implication => {
                self.implications.record(reference);
            }
            AnnotationType::Spec => {
                // do nothing, it's just a reference
            }
        }
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct Stat {
    pub range: u64,
    pub lines: u64,
    cursor: u64,
}

impl Stat {
    fn record(&mut self, reference: &Reference) {
        let start = reference.start() as u64;
        let end = reference.end() as u64;
        let len = end - start.max(self.cursor);
        if len > 0 {
            self.range += len;
            self.lines += 1;
        }
        self.cursor = end;
    }
}
